lib/bolt/inventory/group2.rb

```diff
@@ -0,0 +1,313 @@
+# frozen_string_literal: true
+
+require 'bolt/inventory/group'
+
+module Bolt
+  class Inventory
+    class Group2 < Group
+      attr_accessor :name, :nodes, :aliases, :name_or_alias, :groups, :config, :rest, :facts, :vars, :features
+
+      # THESE are duplicates with the old groups for now.
+      # Regex used to validate group names and target aliases.
+      NAME_REGEX = /\A[a-z0-9_][a-z0-9_-]*\Z/.freeze
+
+      DATA_KEYS = %w[name config facts vars features].freeze
+      NODE_KEYS = DATA_KEYS + ['alias']
+      GROUP_KEYS = DATA_KEYS + %w[groups nodes]
+      CONFIG_KEYS = Bolt::TRANSPORTS.keys.map(&:to_s) + ['transport']
+
+      def initialize(data)
+        @logger = Logging.logger[self]
+
+        raise ValidationError.new("Expected group to be a Hash, not #{data.class}", nil) unless data.is_a?(Hash)
+        raise ValidationError.new("Group does not have a name", nil) unless data.key?('name')
+
+        @name = data['name']
+        raise ValidationError.new("Group name must be a String, not #{@name.inspect}", nil) unless @name.is_a?(String)
+        raise ValidationError.new("Invalid group name #{@name}", @name) unless @name =~ NAME_REGEX
+
+        unless (unexpected_keys = data.keys - GROUP_KEYS).empty?
+          msg = "Found unexpected key(s) #{unexpected_keys.join(', ')} in group #{@name}"
+          @logger.warn(msg)
+        end
+
+        @vars = fetch_value(data, 'vars', Hash)
+        @facts = fetch_value(data, 'facts', Hash)
+        @features = fetch_value(data, 'features', Array)
+        @config = fetch_value(data, 'config', Hash)
+
+        unless (unexpected_keys = @config.keys - CONFIG_KEYS).empty?
+          msg = "Found unexpected key(s) #{unexpected_keys.join(', ')} in config for group #{@name}"
+          @logger.warn(msg)
+        end
+
+        nodes = fetch_value(data, 'nodes', Array)
+        groups = fetch_value(data, 'groups', Array)
+
+        @nodes = {}
+        @aliases = {}
+        nodes.reject { |node| node.is_a?(String) }.each do |node|
+          unless node.is_a?(Hash)
+            raise ValidationError.new("Node entry must be a String or Hash, not #{node.class}", @name)
+          end
+
+          node['name'] ||= node['uri']
+          name = node['name']
```

Comment:
Nit: rename to `node_name` or something to avoid conflict with the group name `@name`